### PR TITLE
Flair!

### PR DESCRIPTION
**Pull request by lukebarnard1**
Merge https://github.com/matrix-org/matrix-react-sdk/pull/1328 into `develop` and then `develop` into `luke/feature-flair`.

Implementation of basic Flair:
 - doesn't change appearance with > 1 groups
 - doesn't do cache-busting

2017-08-30-111447_550x69_scrot

### CODE REVIEW

File: src/components/views/elements/Flair.js [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        userGroups[userId] = groups;
+        // TODO: Reset cache at this point
+        return userGroups[userId];
+    }).catch((err) => {
+        throw err;
+    }).finally(() => {
+        delete usersPending[userId];
+    });
+
+    if (debounceTimeoutID) clearTimeout(debounceTimeoutID);
```
[the remark this answers is not in the document]
Comment (lukebarnard1) [reply src/components/views/elements/Flair.js]: >if you re-set it each time then you could potentially never actually do the request if you always get another lookup every 199ms.

I think you're suggesting polling every 200ms if we have any users pending. Given scroll speeds and the rate at which people might show up in the timeline (i.e. the rate at which we add users to users pending), we could easily end up never doing a bulk request and only ever asking for one user at a time.

File: src/components/views/elements/Flair.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        this.state = {
+            avatarUrls: [],
+        };
+    }
+
+    componentWillUnmount() {
+        this._unmounted = true;
+    }
+
+    async componentWillMount() {
```
[the remark this answers is not in the document]
Comment (lukebarnard1) [reply src/components/views/elements/Flair.js]: oh, I forgot to get rid of that during refactor

File: src/components/views/elements/Flair.js [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    }
+
+    async _generateAvatars() {
+        let groups;
+        try {
+            groups = await getPublicGroupsCached(this.context.matrixClient, this.props.userId);
+        } catch (err) {
+            console.error('Could not get groups for user', this.props.userId, err);
+        }
+        const avatarUrls = await this._getAvatarUrls(groups);
```
[the remark this answers is not in the document]
Comment (lukebarnard1) [reply src/components/views/elements/Flair.js]: oh. yeah that's true, I'll have it return if !groups

File: src/components/views/elements/Flair.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    }
+}
+
+Flair.propTypes = {
+    userId: PropTypes.string,
+};
+
+Flair.contextTypes = {
+    matrixClient: React.PropTypes.instanceOf(MatrixClient).isRequired,
+};
```
[the remark this answers is not in the document]
Comment (lukebarnard1) [reply src/components/views/elements/Flair.js]: This is a class so I couldn't use it and didn't really feel like withMatrixClient would give much value anyway?

[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/elements/Flair.js]: Ah, so yes, withMatrixClient is not compatible with ES6 React Classes, but then when it was written this was what we were saying to use for new code - I thought it still was, if not we should update withMatrixClient to be compatible with our current code style.

That said, if it offers no utility we should get rid of it. Some people using it and some not based on personal preference is not really a way to a sane, consistent codebase.

File: src/components/views/elements/Flair.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/elements/Flair.js]: Well, if the assumption is that sometimes lookups will be more than 200ms apart, your version will still send a separate request for each, the difference will be if =, say, 3 lookups arrive 101ms apart, in which case your version would send one request and my suggestion would send two.

Either way, if there's reasoning like this behind it it's probably worth a comment.

**Comment by lukebarnard1** [comment]
@richvdh - genuine test flake for you https://travis-ci.org/matrix-org/matrix-react-sdk/builds/277320766#L1093

**Pull request merged**